<h1 style="text-align: center;">Week 1 - Probability Models, Conditional Probability and Baye's Theorem</h1>

# Fundamental Concepts 

**Axioms of Probability** 
1. P(A) >= 0, for every event A. 
2. The probability of entire sample space is 1, P($\Omega$) = 1 
3. If A and B are two disjoint events, then P(A $\cup$ B) = P(A) + P(B) 

--- 

**Conditional Probability** 
- Probability of event A given B, written as P(A|B) is - 
$$P(A|B) = \frac{P(A \cap B)}{P(B)}$$

--- 

**Total Probability Theorem**  
$$P(B) = P(B \cap A_1) + P(B \cap A_2) + ... + P(B \cap A_n)$$ 
$$P(B) = P(A_1).P(B|A_1) + P(A_2).P(B|A_2) + ... + P(A_n).P(B|A_n)$$  

--- 

**Bayes' Theorem** 
$$P(A_i|B) = \frac{P(A_i) \cdot P(B|A_i)}{P(B)}$$

$$P(A_i|B) = \frac{P(A_i) \cdot P(B|A_i)}{P(A_1) \cdot P(B|A_1) + (A_2) \cdot P(B|A_2) + ... + (A_n) \cdot P(B|A_n)}$$

---
---

## Monty Hall Problem Simulation

**Problem Statement** 

- There are 3 doors: 1 hides a car, 2 hide goats. You pick a door, the host (who knows what's behind each door) opens a different door revealing a goat, then you decide: stick with your original pick or switch to the remaining door.


In [1]:
import random

In [2]:
def place_a_car():
    n = random.randint(1,3)
    return n 

In [3]:
def user_picks_a_door():
    door = int(input("Pick a door (1, 2, or 3): "))
    return door 

In [4]:
def reveal_a_door(door_picked, car_door):
    door = [1,2,3]
    if door_picked == car_door:
        revealed_door = random.choice([d for d in door if d!= door_picked]) 
    else:
        revealed_door = sum(door) - door_picked - car_door 
    
    return revealed_door

In [5]:
def simulate_one_game():
    car_door = place_a_car()
    user_pick = user_picks_a_door()
    revealed_door = reveal_a_door(user_pick, car_door)
   
    strategy = input(f"Door {revealed_door} has a goat behind it. Do you want to switch or stay?")
    if strategy == "switch":
        user_pick =  6 - user_pick - revealed_door 
    
    if car_door == user_pick:
        print("You won the car")  
    else:
        print("Unlucky, try next time")     

In [7]:
simulate_one_game()

Pick a door (1, 2, or 3): 1
Door 2 has a goat behind it. Do you want to switch or stay?stay
Unlucky, try next time


In [8]:
# let's do a number of simulation and see which strategy works better. 
# istead of taking user inputs we will be randomly picking the doors for each trial while the strategy remains the same
# we will calculate the number of wins for each strategy

In [9]:
def simulate_multiple_trials(n_trials, strategy): 
    
    wins = 0 
    
    for i in range(n_trials):
        user_pick = random.randint(1,3)
        car_door = random.randint(1,3)
        revealed_door = reveal_a_door(user_pick,car_door)
        
        if strategy == "switch":
            user_pick =  6 - user_pick - revealed_door 
    
        if car_door == user_pick:
            wins += 1 
    

    return f"The win rate is {round(wins/n_trials * 100,2)}"

In [10]:
simulate_multiple_trials(1000000, "switch")

'The win rate is 66.64'

In [11]:
simulate_multiple_trials(1000000,'stay')

'The win rate is 33.38'

<span style="color: blue;">
<b>From the simulations it can be seen that the win rate is 66.6% when the user switches and 33.3% when the user chose to stay.</b>

</span>

---
---

#### The Mathematical Solution

**Example:** User picks door 1 and the host reveals door 2 having a goat (event $B$).

Given that door 2 has a goat, we answer the following —

**What is the conditional probability that door 1 has the car?** $P(A|B)$

---

**Defining Events:**

- Event $A$ — door 1 has car, $P(A) = \frac{1}{3}$
- Event $A^c$ — door 1 does not have car, $P(A^c) = \frac{2}{3}$
- $P(B|A)$ — probability door 2 has a goat, given door 1 has car $= 1$
- $P(B|A^c)$ — probability door 2 has a goat, given door 1 does not have car $= 1$

---

**Applying Bayes' Theorem:**

$$P(A|B) = \frac{P(B|A) \cdot P(A)}{P(B|A) \cdot P(A) + P(B|A^c) \cdot P(A^c)}$$

$$P(A|B) = \frac{1 \cdot \frac{1}{3}}{1 \cdot \frac{1}{3} + 1 \cdot \frac{2}{3}}$$

$$P(A|B) = \frac{1}{3}$$

**Not switching leads to a winning chance of 33.33%.**

---

<span style="color: blue;"><b>NOTE —  When door 1 has a goat (i.e., AcA^c
Ac), the car is behind door 2 or door 3. The host *must* open a goat door that isn't door 1, so he is **forced** to open door 2 (the only remaining goat door) — hence $P(B∣A^c)=1$ </b></span>

---
---

## Medical Test - Python Simulation

**Problem Statement** 

A disease affects 1% of the population. A test for this disease is 99% accurate if you have the disease (it correctly returns positive 99% of the time). The test is 95% accurate if you don't have the disease (it correctly returns negative 95% of the time).
You just tested positive. What is the probability that you actually have the disease?

In [12]:
import numpy as np

In [13]:
#let's say the population is 1000_000. Numbering each person.  
population = list(range(1,1000_000))

# randomly chosing 1% users who have the disease

persons_with_disease = np.random.choice(population, 10000, replace = False)
persons_without_disease = list(set(population) - set(persons_with_disease))

In [14]:
persons_with_disease

array([917531, 450630, 227716, ..., 220334, 716972, 939178])

In [15]:
# randomly chosing the 99% 0f users for whom the test result return positive
count_1 = int(0.99 * len(persons_with_disease))
persons_with_disease_tested_positive = np.random.choice(persons_with_disease,count_1, replace = False)

In [16]:
# randomly chosing 5% people for whom the test returned positive even though they don't have the disease 
persons_without_disease_tested_positive = np.random.choice(persons_without_disease, int(0.05* len(persons_without_disease)), replace = False)

In [17]:
persons_tested_positive = np.concatenate((persons_with_disease_tested_positive, persons_without_disease_tested_positive))

In [18]:
#chosing a random person from the list of persons who tested positive 

subject = random.choice(persons_tested_positive)

In [19]:
subject

180274

In [20]:
subject in persons_with_disease

False

In [21]:
#simulating for all the subject

x = 0 
for subject in persons_tested_positive: 
    if subject in persons_with_disease:
        x+=1 

print(f"if you have been tested positive, there are {round(x/len(persons_tested_positive) * 100,2 )} chances that you have the disease")
        

if you have been tested positive, there are 16.67 chances that you have the disease
